File: in_Python/1019 Next Greater Node In Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def nextLargerNodes(self, head):
        """
        :type head: ListNode
        :rtype: List[int]
        """
        res = []
        while head != None:
            res.append(head.val)
            head = head.next
        st = []
        result = []
        for i in res: result.append(0)
        for i in range(len(res)):
            while len(st) != 0 and res[st[-1]] < res[i]:
                result[st[-1]] = res[i]
                st.pop()
            st.append(i)

        return result

File: in_Java/1019 Next Greater Node In Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public int[] nextLargerNodes(ListNode head) {
        int leng = 0;
        ListNode d = head;
        while(d != null){
            d = d.next;
            leng++;
        }
        int[] arr = new int[leng];
        int id = 0;
        ListNode dd = head;
        while(dd != null){
            arr[id++] = dd.val;
            dd = dd.next;
        }
        int[] res = new int[leng];
        Stack<Integer> st = new Stack<>();
        for(int i = 0; i < leng; i++){
            if(st.isEmpty() || arr[i] <= arr[st.peek()]){
                st.push(i);
            }else{
                while(!st.isEmpty() && arr[i] > arr[st.peek()]){
                    res[st.peek()] = arr[i];
                    st.pop();
                }
                st.push(i);
            }
        }
        return res;
    }
}